prime-router/src/test/kotlin/azure/TokenFunctionTests.kt

```diff
@@ -278,6 +301,91 @@ class TokenFunctionTests {
+        httpRequestMessage.parameters.put("scope", validScope)
+        // Invoke
+        var response = TokenFunction(UnitTestUtils.simpleMetadata).token(httpRequestMessage)
+        // Verify
+        assertThat(response.getStatus()).isEqualTo(HttpStatus.OK)
+    }
+
+    // TODO https://github.com/CDCgov/prime-reportstream/issues/8659
+    // This and the following test can be removed after all keys associated with senders have been moved to
+    // the organization.  For now these tests cover the possibility that keys for the same scope might exist on the
+    // organization and sender; in that case the organization keys are considered first.
+    @Test
+    fun `Test success when sender key is broken, but organization key is not`() {
+        mockkConstructor(Server2ServerAuthentication::class)
+        every {
+            anyConstructed<Server2ServerAuthentication>().createAccessToken(any(), any(), any())
+        } returns AccessToken(
+            "test",
+            "test",
+            "test",
+            10,
+            10,
+            "test"
+        )
+
+        settings.senderStore.put(sender.fullName, CovidSender(sender, validScope, badJwk))
+        settings.organizationStore.put(organization.name, Organization(organization, validScope, jwk))
+
+        var httpRequestMessage = MockHttpRequestMessage()
+        httpRequestMessage.parameters.put("client_assertion", token)
+        httpRequestMessage.parameters.put("scope", validScope)
+        // Invoke
+        var response = TokenFunction(UnitTestUtils.simpleMetadata).token(httpRequestMessage)
+        // Verify
+        assertThat(response.getStatus()).isEqualTo(HttpStatus.OK)
+    }
+
+    @Test
+    fun `Test success when organization key is broken, but sender key is not`() {
+        mockkConstructor(Server2ServerAuthentication::class)
+        every {
+            anyConstructed<Server2ServerAuthentication>().createAccessToken(any(), any(), any())
+        } returns AccessToken(
+            "test",
+            "test",
+            "test",
+            10,
+            10,
+            "test"
+        )
+
+        settings.senderStore.put(sender.fullName, CovidSender(sender, validScope, jwk))
+        settings.organizationStore.put(organization.name, Organization(organization, validScope, badJwk))
+
+        var httpRequestMessage = MockHttpRequestMessage()
+        httpRequestMessage.parameters.put("client_assertion", token)
+        httpRequestMessage.parameters.put("scope", validScope)
+        // Invoke
+        var response = TokenFunction(UnitTestUtils.simpleMetadata).token(httpRequestMessage)
+        // Verify
+        // This fails because the current logic short circuits if the JWK scope matches the requested scope; the
+        // organization JwkSet has the correct scope, but the key is not valid, so not key is returned.
+        // See: Server2ServerAuthentication.kt#225
+        assertThat(response.getStatus()).isEqualTo(HttpStatus.UNAUTHORIZED)
```

Comment:
It was unclear whether the logic in the server2server code needs to get updated, but this assertion is accurate based on current implementation, since the code will stop processing the remaining keys after one matches the requested scope regardless if the key can ultimately be used.

Reply:
Yes, my vague memory is that may be a flaw in the server2server auth code - it only tries the first key, not all the possible keys.

Reply:
Ok... I'll be creating a bug ticket for this

Reply:
https://github.com/CDCgov/prime-reportstream/issues/8670